File: sdk/rust/src/artifact/rust_src.rs
use crate::{
    api::artifact::ArtifactSystem::{Aarch64Darwin, Aarch64Linux, X8664Darwin, X8664Linux},
    artifact::{rust_toolchain, step, Artifact, ArtifactSource},
    context::ConfigContext,
};
use anyhow::Result;

#[derive(Default)]
pub struct RustSrc {}

impl RustSrc {
    pub fn new() -> Self {
        Self::default()
    }

    pub async fn build(self, context: &mut ConfigContext) -> Result<String> {
        let name = "rust-src";
        let source_version = rust_toolchain::version();
        let source_path =
            format!("https://static.rust-lang.org/dist/rust-src-{source_version}.tar.gz");

        let source = ArtifactSource::new(name, source_path.as_str()).build();

        let step_script = format!(
            "cp -prv \"./source/{name}/{name}-{source_version}/{name}/.\" \"$VORPAL_OUTPUT\""
        );
        let steps = vec![step::shell(context, vec![], vec![], step_script, vec![]).await?];
        let systems = vec![Aarch64Darwin, Aarch64Linux, X8664Darwin, X8664Linux];

        Artifact::new(name, steps, systems)
            .with_sources(vec![source])
            .build(context)
            .await
    }
}

File: sdk/typescript/src/artifact/rust_src.ts
import type { ConfigContext } from "../context.js";
import { Artifact, ArtifactSource } from "../artifact.js";
import { shell } from "./step.js";
import { RUST_TOOLCHAIN_VERSION } from "./rust_toolchain.js";
import { SYSTEMS } from "../system.js";

/**
 * Builder for the RustSrc artifact.
 *
 * Mirrors Rust `RustSrc` struct in `sdk/rust/src/artifact/rust_src.rs`.
 * Downloads and extracts the Rust source from static.rust-lang.org.
 * Note: rust-src is platform-independent (no target in URL).
 */
export class RustSrc {
  async build(context: ConfigContext): Promise<string> {
    const name = "rust-src";
    const sourceVersion = RUST_TOOLCHAIN_VERSION;
    const sourcePath = `https://sdk.vorpal.build/source/rust-src-${sourceVersion}.tar.gz`;

    const source = new ArtifactSource(name, sourcePath).build();

    const stepScript = `cp -pr "./source/${name}/${name}-${sourceVersion}/${name}/." "$VORPAL_OUTPUT"`;
    const steps = [await shell(context, [], [], stepScript, [])];

    return new Artifact(name, steps, SYSTEMS)
      .withSources([source])
      .build(context);
  }
}
